tests/unittests/MatrixRotationTest.cpp

```diff
@@ -0,0 +1,568 @@
+
+  auto addToMatrixVector = [&](const Matrix3x3 &curMatrix) {
+    out << "  ";
+    // The print function is expected to print something compatible with
+    // a static initialization of float[3][3].
+    curMatrix.print(out);
+    out << ",\n";
+
+    matrixData.push_back(curMatrix);
+    ++currentIdx;
+  };
+
+  for (int i = 0; i < size; ++i) {
+    int baseIdx = currentIdx;
+    assert(currentIdx == matrixData.size() && "Index out-of-sync");
+    out << "  /* Base matrix " << baseIdx << " */\n";
+    Matrix3x3 curMatrix(-range, range);
+
+    addToMatrixVector(curMatrix);
+
+    // Generate a bunch of rotation from the base matrix.
+    for (int currentRotation = 0, maxRotations = nextRandInt(1, 8);
+         currentRotation != maxRotations; ++currentRotation) {
+      int rotationSteps = nextRandInt(0, 7);
+      Matrix3x3 rotatedMatrix = curMatrix.rotate(rotationSteps);
+      rotationRelationshipData.emplace_back(
+          RotationRelationship{baseIdx, currentIdx, rotationSteps});
+
+      out << "  /* Rotation matrix " << currentIdx << " == " << baseIdx
+          << " rotated " << rotationSteps << " */\n";
+      addToMatrixVector(rotatedMatrix);
+    }
+  }
+
+  // Generate pairs without rotations.
+  for (int outOfSync = 0, maxOutOfSync = nextRandInt(size / 2, size * 10);
+       outOfSync != maxOutOfSync; ++outOfSync) {
+    // Choose two matrices randomly.
+    // If one is not a rotation of the other, record that pair.
+    // Otherwise create a new matrix that breaks the rotation relationship
+    // and record that pair.
+    assert(currentIdx == matrixData.size() && "Index out-of-sync");
+    int matrixIdx1 = nextRandInt(0, currentIdx - 1);
+    int matrixIdx2 = nextRandInt(0, currentIdx - 1);
+    const Matrix3x3 *M1 = &matrixData[matrixIdx1];
+    const Matrix3x3 &M2 = matrixData[matrixIdx2];
+
+    if (M1->clockwiseRotationStepsToGet(M2) != -1) {
+      Matrix3x3 tmp = M1->breakRotation();
+      out << "  /* Out-of-sync matrix " << currentIdx << " broke " << matrixIdx1
+          << " */\n";
+      matrixIdx1 = currentIdx;
+      addToMatrixVector(tmp);
+    }
+    M1 = &matrixData[matrixIdx1];
+    assert(M1->clockwiseRotationStepsToGet(M2) == -1);
+    rotationRelationshipData.emplace_back(
+        RotationRelationship{matrixIdx1, matrixIdx2, -1});
+  }
+
+  out << "}; // End of " << varName << ".\n\n\n";
+
+  // Now print the relationships.
+  llvm::Twine relationshipVarName =
+      llvm::Twine(varName) + llvm::Twine("RelationshipData");
+  out << "static const RotationRelationship " << relationshipVarName
+      << "[] = {\n";
+
+  for (const RotationRelationship &rel : rotationRelationshipData) {
+    out << "  {" << rel.matrixIdxA_ << ',' << rel.matrixIdxB_ << ','
+        << rel.clockwiseRotationStepsAToB_ << "},\n";
+  }
+
+  out << "}; // End of " << relationshipVarName << ".\n";
+}
+
+
+// --------------------- Tests for Matrix3x3 start here ----------------------
+static float m1[3][3] = {
```

Comment:
nit: you can make all those std::array and make ctor ```Matrix3x3(const float m[3][3]) ``` to get an array as input param.